builder.go

```diff
@@ -181,73 +128,185 @@ func (b *Builder) Build() (*BuildMetadata, error) {
 }
 
-func (p BuildpackPlan) has(entry BuildPlanEntry) bool {
-	for _, buildEntry := range p.Entries {
-		for _, req := range entry.Requires {
-			if req.Name == buildEntry.Name {
-				return true
-			}
+func (b *BuildpackTOML) runBuildCmd(bpLayersDir, bpPlanPath string, config BuildConfig) error {
+	cmd := exec.Command(
+		filepath.Join(b.Path, "bin", "build"),
+		bpLayersDir,
+		config.PlatformDir,
+		bpPlanPath,
+	)
+	cmd.Dir = config.AppDir
+	cmd.Stdout = config.Out
+	cmd.Stderr = config.Err
+
+	var err error
+	if b.Buildpack.ClearEnv {
+		cmd.Env = config.Env.List()
+	} else {
+		cmd.Env, err = config.Env.WithPlatform(config.PlatformDir)
+		if err != nil {
+			return err
 		}
 	}
-	return false
+	cmd.Env = append(cmd.Env, EnvBuildpackDir+"="+b.Path)
+
+	if err := cmd.Run(); err != nil {
+		return NewLifecycleError(err, ErrTypeBuildpack)
+	}
+	return nil
 }
 
-func (bom *BOMEntry) convertMetadataToVersion() error {
-	if version, ok := bom.Metadata["version"]; ok {
-		metadataVersion := fmt.Sprintf("%v", version)
-		if bom.Version != "" && bom.Version != metadataVersion {
-			return errors.New("top level version does not match metadata version")
+func preparePaths(bpID string, foundPlan BuildpackPlan, layersDir, planDir string) (string, string, error) {
+	bpDirName := launch.EscapeID(bpID)
+	bpLayersDir := filepath.Join(layersDir, bpDirName)
+	bpPlanDir := filepath.Join(planDir, bpDirName)
+	if err := os.MkdirAll(bpLayersDir, 0777); err != nil {
+		return "", "", err
+	}
+	if err := os.MkdirAll(bpPlanDir, 0777); err != nil {
+		return "", "", err
+	}
+	bpPlanPath := filepath.Join(bpPlanDir, "plan.toml")
+	if err := WriteTOML(bpPlanPath, foundPlan); err != nil {
+		return "", "", err
+	}
+
+	return bpLayersDir, bpPlanPath, nil
+}
+
+func readOutputFiles(bp Buildpack, bpLayersDir, bpPlanPath string, bpPlanIn BuildpackPlan) (BuildResult, error) {
+	// read buildpack plan
+	var bpPlanOut BuildpackPlan
+	if _, err := toml.DecodeFile(bpPlanPath, &bpPlanOut); err != nil {
+		return BuildResult{}, err
+	}
+	// read build.toml
+	var bpBuild BuildTOML
+	if api.MustParse(bp.API).Compare(api.MustParse("0.5")) >= 0 { // bp.API >= 0.5
+		buildPath := filepath.Join(bpLayersDir, "build.toml")
+		if _, err := toml.DecodeFile(buildPath, &bpBuild); err != nil && !os.IsNotExist(err) {
+			return BuildResult{}, err
+		}
+		if err := validateBuildTOML(bpBuild, bpPlanIn); err != nil {
+			return BuildResult{}, err
 		}
-		bom.Version = metadataVersion
 	}
-	return nil
+	// read launch.toml
+	var launch LaunchTOML
+	launchPath := filepath.Join(bpLayersDir, "launch.toml")
+	if _, err := toml.DecodeFile(launchPath, &launch); err != nil && !os.IsNotExist(err) {
+		return BuildResult{}, err
+	}
+	// setup result
+	var br BuildResult
+	if api.MustParse(bp.API).Compare(api.MustParse("0.5")) < 0 { // bp.API <= 0.4
+		br.BOM = withBuildpack(bp, bpPlanOut.toBOM())
+		br.Met = bpPlanOut.Entries
+	} else {
+		br.BOM = withBuildpack(bp, launch.BOM)
```

Comment:
We should be validating that the top-level version is not set in the launch.BOM entries. We don't want to carry-over the deprecated usage to this new file.